### PR TITLE
[Enhancement] - Newsletter Screen updates

### PR DESCRIPTION
**Pull request by Rcureton**
## Why
- Continuing the implementation of the new Settings 

## What
- Implemented the `NewsletterActivity` and `NewsletterViewModel` using Kotlin as well as the tests.

## Todo
 - Need to implement the new methods to the `SettingsBody` to incorporate the new networking calls for additional Newsletters.

## See
screenshot_1531857832

### CODE REVIEW

File: app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt [hunk 1]
```diff
@@ -1,13 +1,83 @@
...
+                .compose(bindToLifecycle())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe { ViewUtils.showToast(this, getString(R.string.profile_settings_error)) }
+
+        this.viewModel.outputs.showOptInPrompt()
+                .compose(bindToLifecycle())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(this::showOptInPrompt)
+
+        binding.apply {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt]: Instance variables should be referenced by using `this`. Going forward I'll comment "this" where I see it's missing.

File: app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt [hunk 2]
```diff
@@ -1,13 +1,83 @@
...
+        }
+    }
+
+    private fun displayPreferences(@NonNull user: User) {
+        SwitchCompatUtils.setCheckedWithoutAnimation(this.binding.happeningSwitch, isTrue(user.happeningNewsletter()))
+        SwitchCompatUtils.setCheckedWithoutAnimation(this.binding.newsEventsSwitch, isTrue(user.promoNewsletter()))
+        SwitchCompatUtils.setCheckedWithoutAnimation(this.binding.projectsWeLoveSwitch, isTrue(user.weeklyNewsletter()))
+    }
+
+    private fun newsletterString(newsletter: Newsletter): String? {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt]: Should `newsletter` also be `@NonNull`?

File: app/src/main/java/com/kickstarter/viewmodels/HelpNewViewModel.kt [hunk 3]
```diff
@@ -18,6 +18,7 @@ import java.util.*
 class HelpNewViewModel(private val context: Context, private val environment: Environment) : BaseObservable() {
 
     fun contactClick() {
+        environment.koala().trackContactEmailClicked()
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/HelpNewViewModel.kt]: `this`

File: app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        fun sendPromoNewsletter(checked: Boolean)
+
+        /** Call when the user toggles the Projects We Love newsletter switch.  */
+        fun sendWeeklyNewsletter(checked: Boolean)
+
+    }
+
+    interface Outputs {
+        /** Emits user containing settings state. */
+        fun user(): Observable<User>
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt]: Let's alpha these outputs.

File: app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+        /** Show a dialog to inform the user that their newsletter subscription must be confirmed via email.  */
+         fun showOptInPrompt(): Observable<Newsletter>
+    }
+
+    interface Errors {
+        fun unableToSavePreferenceError(): Observable<String>
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt]: Let's add a comment description this error.

File: app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+        private val client = environment.apiClient()
+        private val currentUser = environment.currentUser()
+        private val newsletterInput = PublishSubject.create<Pair<Boolean, Newsletter>>()
+        private val showOptInPrompt = PublishSubject.create<Newsletter>()
+        private val userInput = PublishSubject.create<User>()
+        private val updateSuccess = PublishSubject.create<Void>()
+        private val userOutput = BehaviorSubject.create<User>()
+
+
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt]: let's delete this nl

File: app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            this.userInput
+                    .concatMap { updateSettings(it) }
+                    .compose(bindToLifecycle())
+                    .subscribe { success(it) }
+
+            this.userInput
+                    .compose(bindToLifecycle())
+                    .subscribe(this.userOutput)
+
+
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt]: let's delete this nl

File: app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            this.userOutput
+                    .window(2, 1)
+                    .flatMap<List<User>>({ it.toList() })
+                    .map<User>({ ListUtils.first(it) })
+                    .compose<User>(takeWhen<User, Throwable>(this.unableToSavePreferenceError))
+                    .compose(bindToLifecycle())
+                    .subscribe(this.userOutput)
+
+
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt]: nl

File: app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+            this.newsletterInput
+                    .map { bs -> bs.first }
+                    .compose(bindToLifecycle())
+                    .subscribe(this.koala::trackNewsletterToggle)
+        }
+
+        private fun requiresDoubleOptIn(user: User, checked: Boolean) = UserUtils.isLocationGermany(user) && checked
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/NewsletterViewModel.kt]: Let's move all private helper methods to the bottom of the class.

File: app/src/main/java/com/kickstarter/viewmodels/SettingsNewViewModel.kt [hunk 10]
```diff
@@ -20,6 +20,7 @@ class SettingsNewViewModel(private val context: Context, private val environment
 
     fun logoutClick() {
         environment.logout().execute()
+        environment.koala().trackLogout()
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/SettingsNewViewModel.kt]: `this`

File: app/src/test/java/com/kickstarter/viewmodels/NewsletterViewModelTest.kt [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+        this.vm.inputs.sendHappeningNewsletter(true)
+        this.koalaTest.assertValues("Newsletter Subscribe")
+        this.currentUserTest.assertValues(user, user.toBuilder().happeningNewsletter(true).build())
+
+        this.vm.inputs.sendHappeningNewsletter(false)
+        this.koalaTest.assertValues("Newsletter Subscribe", "Newsletter Unsubscribe")
+        this.currentUserTest.assertValues(user, user.toBuilder().happeningNewsletter(true).build(), user)
+
+        showOptInPromptTest.assertNoValues()
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/NewsletterViewModelTest.kt]: `this`

File: app/src/main/res/layout/activity_newsletter.xml [hunk 12]
```diff
@@ -1,13 +1,138 @@
...
+              android:id="@+id/news_events_switch"
+              style="@style/NewsLetterSwitch" />
+
+          </LinearLayout>
+        </LinearLayout>
+
+        <TextView
+          style="@style/NewsLetterTextView"
+          android:text="@string/News_events" />
+
```
Comment (eoji) [inline app/src/main/res/layout/activity_newsletter.xml]: I think we're missing the Games newsletter.

File: app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt [hunk 13]
```diff
@@ -1,13 +1,83 @@
...
+        return when (newsletter) {
+            Newsletter.HAPPENING -> getString(R.string.profile_settings_newsletter_happening)
+            Newsletter.PROMO -> getString(R.string.profile_settings_newsletter_promo)
+            Newsletter.WEEKLY -> getString(R.string.profile_settings_newsletter_weekly)
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt]: I think we're missing the Games newsletter.

File: app/src/test/java/com/kickstarter/viewmodels/NewsletterViewModelTest.kt [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        this.koalaTest.assertValues("Newsletter Subscribe")
+        this.currentUserTest.assertValues(user, user.toBuilder().weeklyNewsletter(true).build())
+
+        this.vm.inputs.sendWeeklyNewsletter(false)
+        this.koalaTest.assertValues("Newsletter Subscribe", "Newsletter Unsubscribe")
+        this.currentUserTest.assertValues(user, user.toBuilder().weeklyNewsletter(true).build(), user)
+
+        this.showOptInPromptTest.assertNoValues()
+    }
+}
```
Comment (eoji) [inline app/src/test/java/com/kickstarter/viewmodels/NewsletterViewModelTest.kt]: nl

File: app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt (hunk 13 above)
Comment (Rcureton) [reply app/src/main/java/com/kickstarter/ui/activities/NewsletterActivity.kt]: Yes I added it in this second PR I have coming to add the rest of the Newsletter switches. At the time when I started this PR Danny didn't have Games in the wireframes, so I have it in the next PR wired up to go.

File: app/src/main/res/layout/activity_newsletter.xml (hunk 12 above)
Comment (Rcureton) [reply app/src/main/res/layout/activity_newsletter.xml]: See comment below.

**Pull request merged**